src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackDecoder.cs

```diff
@@ -112,256 +110,278 @@ internal HPackDecoder(int maxDynamicTableSize, int maxResponseHeadersLength, Dyn
             _maxDynamicTableSize = maxDynamicTableSize;
             _maxResponseHeadersLength = maxResponseHeadersLength;
             _dynamicTable = dynamicTable;
+
+            _stringOctets = new byte[maxResponseHeadersLength];
+            _headerNameOctets = new byte[maxResponseHeadersLength];
+            _headerValueOctets = new byte[maxResponseHeadersLength];
+        }
+
+        public void Decode(in ReadOnlySequence<byte> data, bool endHeaders, IHttpHeadersHandler handler)
+        {
+            foreach (ReadOnlyMemory<byte> segment in data)
+            {
+                ReadOnlySpan<byte> span = segment.Span;
+                for (int i = 0; i < span.Length; i++)
+                {
+                    OnByte(span[i], handler);
+                }
+            }
+
+            CheckIncompleteHeaderBlock(endHeaders);
         }
 
-        public void Decode(ReadOnlySpan<byte> data, bool endHeaders, HeaderCallback onHeader, object onHeaderState)
+        public void Decode(in ReadOnlySpan<byte> data, bool endHeaders, IHttpHeadersHandler handler)
         {
             for (int i = 0; i < data.Length; i++)
             {
-                byte b = data[i];
+                OnByte(data[i], handler);
```

Comment:
Have you measured the impact of this on throughput?  This is adding a non-inlineable function call per byte.

Reply:
I will.

Reply:
```
|                         Method |     Mean |     Error |    StdDev |
|------------------------------- |---------:|----------:|----------:|
| BenchmarkDecodeWithoutFunction | 2.151 us | 0.0172 us | 0.0153 us |
|    BenchmarkDecodeWithFunction | 2.205 us | 0.0429 us | 0.0511 us |
```

Small difference. Refactor was straight forward though so I went ahead and did it.